Panache REST guide (#9254)
quarkusio/quarkus: docs/src/main/asciidoc/panache-rest.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////

= Quarkus - Generating CRUD resources with Panache REST

include::./attributes.adoc[]
:extension-status: experimental

> A lot of web applications are monotonous CRUD applications with boring to write REST APIs.

Reviewer: ```suggestion
A lot of web applications are monotonous CRUD applications with REST APIs that are tedious to write.
```

> To ease this task Panache REST can generate the basic CRUD endpoints for your entities and repositories.

Reviewer: ```suggestion
To streamline this task, Panache REST can generate the basic CRUD endpoints for your entities and repositories.
```

> While this extension is still experimental and provides a limited feature set, we hope to get an early feedback for it.

Reviewer: Perhaps this should be in a note?

Author: There is a generic note about the experimental feature already. Do you think I should add another one?

Reviewer: Oh OK. If it's used elsewhere as well, let's keep it as it is

** One of the RestEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)

[source,xml]
----
<dependencies>
    <!-- Panache REST with Hibernate ORM -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-panache-rest-hibernate-orm</artifactId>
    </dependency>

    <!-- JDBC driver dependencies -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-jdbc-postgresql</artifactId>
    </dependency>

    <!-- JSON serializer dependency -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-resteasy-jackson</artifactId>
    </dependency>
</dependencies>
----

* Implement the Panache entities and/or repositories as explained in the link:hibernate-orm-panache[Panache with Hibernate ORM guide].
* Define the interfaces for generation as explained in the sections below.

== Generating CRUD resources

Panache REST generates CRUD resources based on the interfaces available in your application.
For each entity and repository that you want to generate, provide an interface that extends either `PanacheEntityCrudResource` or `PanacheRepositoryCrudResource` interface.

> _Do not implement these interfaces and don't provide custom methods because they won't be implemented._

Reviewer: ```suggestion
_Do not implement these interfaces and don't provide custom methods because they be ignored._
```

Author: `will be ignored` maybe?

Reviewer: yeah :)

Author: How do you edit string in the PR?

Reviewer: See `6` from https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request

You can, however, override the methods from the extended interface in order to customize them (see the section at the end).

The resources that will be generated will be functionally equivalent for both entities and repositories.

> Only the data will accessed using either one or the other.

Reviewer: This line doesn't seem to make sense

Author: How about `The generated resources will be functionally equivalent for both entities and repositories. Except that the particular data access pattern will be used.`

Reviewer: My counter: 

```
The generated resources will be functionally equivalent for both entities and repositories., the difference being in the particular data access pattern will be used.
```

Author: `The only difference being the particular data access pattern in use.`?

Reviewer: Yeah sorry. I really need to pay more attention to what I actually write as opposed to what I am thinking...

The skeleton of the generated JAX-RS resource will look like this:

> [source,java]

Reviewer: I think we need a little more emphasis that on the fact this is generated - not something the user would write.
Mentioning it because looking at the code initially through me off...

Author: It does say "generated JAX-RS resource". Any suggestion?

Reviewer: Yeah I know, it's just that it's too easy to think you need to write this. 
Perhaps change the name of the class and add a comment on it?

Author: I've changed the sentence above

----
@Path("/entities") // Base bath is a pluralised name of an entity
public class EntityCrudResourceImpl implements EntityCrudResource { // The actual class name is going to be unique
    @GET
    @Produces("application/json")
    @Path("{id}")
    public Entity get(@PathParam("id") ID id){
        // ...
    }

    @GET
    @Produces("application/json")
    public List<Entity> list(){
        // ...
    }

    @Transactional
    @POST
    @Consumes("application/json")
    @Produces("application/json")
    public Response add(Entity entity) {
        // ..
    }

    @Transactional
    @PUT
    @Consumes("application/json")
    @Produces("application/json")
    @Path("{id}")
    public Response update(@PathParam("id") ID id, Entity entity) {
        // ..
    }

    @Transactional
    @DELETE
    @Path("{id}")
    public void delete(@PathParam("id") ID id) {
        // ..
    }
}
----

=== PanacheEntityCrudResource

> If your application has an entity (e.g. `Person`) that extends either `PanacheEntity` or `PanacheEntityBase` class, you could tell Panache REST to generate its CRUD resource with the following interface:

Reviewer: ```suggestion
If your application has an entity (e.g. `Person`) that extends either `PanacheEntity` or `PanacheEntityBase` class, you could instruct Panache REST to generate its CRUD resource with the following interface:
```

[source,java]
----
public interface PersonCrudResource extends PanacheEntityCrudResource<Person, Long> {
}
----

=== PanacheRepositoryCrudResource

> If your application has a simple entity (e.g. `Person`) and a repository (e.g. `PersonRepository`) that implements either `PanacheRepository` or `PanacheRepositoryBase` interface, you could tell Panache REST to generate its CRUD resource with the following interface:

Reviewer: ```suggestion
If your application has a simple entity (e.g. `Person`) and a repository (e.g. `PersonRepository`) that implements either `PanacheRepository` or `PanacheRepositoryBase` interface, you could instruct Panache REST to generate its CRUD resource with the following interface:
```

[source,java]
----
public interface PersonCrudResource extends PanacheRepositoryCrudResource<PersonRepository, Person, Long> {
}
----

== Resource customisation

> Panache REST provides a `@PanacheRestResource` annotation that can be used to customize certain features of the resource.

Author: It's out of scope for this PR but I'm not happy with a name of this annotation (`@PanacheRestResource`). It's very similar to the interface name (`PanacheCrud..Resource`) and might be confusing. Any suggestions on the name change @geoand?

Reviewer: I dunno.... I'm probably the worst developer when it comes to naming...

It can be used on your CRUD resource interface:

[source,java]
----
@PanacheRestResource(hal = true)
public interface PersonCrudResource extends PanacheEntityCrudResource<Person, Long> {
    @PanacheRestResource(exposed = false)
    void delete(Long id);
}
----

=== Available options

> * `hal` - in addition to the standard `application/json` responses, generates additional methods that can return `application/hal+json` responses if requested via `Accept` header.

Reviewer: Perhaps a link to what `hal` is would be nice

Author: I thought about it, but then though it might be too much info. Do you think it would be useful?

Reviewer: I think it would be nice to have

Author: Done

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Generating CRUD resources with Panache REST

include::./attributes.adoc[]
:extension-status: experimental

A lot of web applications are monotonous CRUD applications with REST APIs that are tedious to write.
To streamline this task, Panache REST can generate the basic CRUD endpoints for your entities and repositories.

While this extension is still experimental and provides a limited feature set, we hope to get an early feedback for it.
Currently this extension supports Hibernate ORM with Panache and can generate CRUD resources that work with `application/json` and `application/hal+json` content.

include::./status-include.adoc[]

== Setting up Panache REST

To begin with:

* Add the required dependencies to your `pom.xml`

> ** Panache REST with Hiberante ORM extension (`quarkus-panache-rest-hibernate-orm`)

Reviewer 2: Now that I read it like that, I've my doubts over the order, because originally it was supposed to be `X with Panache`, which means `X with style`.

This makes it looks like `Panache` is a noun and the subject instead of an adjective to make Hibernate ORM/REST stylish.

Perhaps it should be `Hibernate ORM REST with Panache` or `RESTful Hibernate ORM with Panache` or `REST for Hibernate ORM with Panache`?

Reviewer 2: I guess, similarly, this guide (and the extension names) would better be "REST with Panache" and `rest-panache`?

Except this sort-of implies it applies to REST apps that aren't necessarily about data, and we plan to have a Panache layer for RESTEasy one day…

Perhaps `RESTful data with Panache`?

Reviewer 2: Or `CRUD with Panache`?

Reviewer 3: In any case, we need to make the extension name consistent with the naming rules we define here.

Author: Should I create a separate issue to agree on the name there?

Reviewer 2: We can, yes, but this will heavily impact this guide.

Author: I could just update the guide together with the code once we agree on a name.

Reviewer 2: OK fine.

Author: https://github.com/quarkusio/quarkus/issues/9293

** A JDBC driver extension (`quarkus-jdbc-postgresql`, `quarkus-jdbc-h2`, `quarkus-jdbc-mariadb`, ...)
** One of the RestEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)

[source,xml]
----
<dependencies>
    <!-- Panache REST with Hibernate ORM -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-panache-rest-hibernate-orm</artifactId>
    </dependency>

    <!-- JDBC driver dependencies -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-jdbc-postgresql</artifactId>
    </dependency>

    <!-- JSON serializer dependency -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-resteasy-jackson</artifactId>
    </dependency>
</dependencies>
----

> * Implement the Panache entities and/or repositories as explained in the link:hibernate-orm-panache[Panache with Hibernate ORM guide].

Reviewer 2: ```suggestion
* Implement the Panache entities and/or repositories as explained in the link:hibernate-orm-panache[Hibernate ORM with Panache guide].
```

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////

> 

Reviewer 3: Please avoid the newline here, the website is very picky about that.

= Quarkus - Generating CRUD resources with Panache REST

include::./attributes.adoc[]
:extension-status: experimental

A lot of web applications are monotonous CRUD applications with REST APIs that are tedious to write.
To streamline this task, Panache REST can generate the basic CRUD endpoints for your entities and repositories.

While this extension is still experimental and provides a limited feature set, we hope to get an early feedback for it.
Currently this extension supports Hibernate ORM with Panache and can generate CRUD resources that work with `application/json` and `application/hal+json` content.

include::./status-include.adoc[]

== Setting up Panache REST

To begin with:

* Add the required dependencies to your `pom.xml`
** Panache REST with Hiberante ORM extension (`quarkus-panache-rest-hibernate-orm`)
** A JDBC driver extension (`quarkus-jdbc-postgresql`, `quarkus-jdbc-h2`, `quarkus-jdbc-mariadb`, ...)

> ** One of the RestEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)

Reviewer: ```suggestion
** One of the RESTEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)
```

[source,xml]
----
<dependencies>
    <!-- Panache REST with Hibernate ORM -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-panache-rest-hibernate-orm</artifactId>
    </dependency>

    <!-- JDBC driver dependencies -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-jdbc-postgresql</artifactId>
    </dependency>

    <!-- JSON serializer dependency -->
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-resteasy-jackson</artifactId>
    </dependency>
</dependencies>
----

* Implement the Panache entities and/or repositories as explained in the link:hibernate-orm-panache[Panache with Hibernate ORM guide].
* Define the interfaces for generation as explained in the sections below.

== Generating CRUD resources

Panache REST generates CRUD resources based on the interfaces available in your application.

> For each entity and repository that you want to generate, provide an interface that extends either `PanacheEntityCrudResource` or `PanacheRepositoryCrudResource` interface.

Reviewer 3: ```suggestion
For each entity and repository you want to expose as a REST service, provide an interface that extends either `PanacheEntityCrudResource` or `PanacheRepositoryCrudResource` interface.
```
would be more correct if I understand the extension correctly?

_Do not implement these interfaces and don't provide custom methods because they will be ignored._ You can, however, override the methods from the extended interface in order to customize them (see the section at the end).

The generated resources will be functionally equivalent for both entities and repositories.
The only difference being the particular data access pattern in use.

Panache REST will generate a JAX-RS resource class similar to this:

[source,java]
----

> @Path("/entities") // Base bath is a pluralised name of an entity

Reviewer 3: I think this is a mistake. We shouldn't "pluralize" things as there's a good chance we will get some of them wrong and it really doesn't bring anything.

Keep in mind that a lot of people code in their own languages where the plural rules will be completely different.

Also, frankly, I don't think it's better as a plural so why take the risk?

Reviewer 3: /cc @FroMage

Reviewer: If we want plural, perhaps let the user specify it?

Reviewer 3: Just to give an example, in French, we have a lot of specific plural forms (`cheval` -> `chevaux` for instance).

Reviewer 3: Yeah, taking the entity name as is would be the default and you could define a specific path in the annotation?

Author: Would you mind to create an issue for this and assign to me? I'll need to change the extension because it now pluralises everything.
I'll also add a way to override it separately.

Just an FIY Spring Data REST simply adds `s` to every entity name.

Reviewer 3: > Just an FIY Spring Data REST simply adds s to every entity name.

Oh, my. OK I will create an issue.

Reviewer 2: Makes it even more fun with:

```java
@Entity
public class 馬 extends PanacheEntity {
}
```

And the URI ends up being `/馬s`, right? :)

Myself, I've always used `Cheval` for the entity name (singular) and `Chevaux` (plural) for the controller name. And RoR derives the URIs from the controller names, so it would be `/chevaux` correctly.

I just suggest we advertise the same convention, it solves many issues. And we default to the controller name for the URI, instead of the entity name.

Naturally we should also allow users to override it.

Author: > And the URI ends up being `/馬s`, right? :)

I would think so. At least for a simple `people` I got a path `/peoples` when i tried it last time.

> I just suggest we advertise the same convention, it solves many issues. And we default to the controller name for the URI, instead of the entity name.

Extracting a name from a controller will be trickier because a user can name his classes differently. We could recommend a convention but who knows if they'll follow it.
With an entity we could simply take an its name, make it all lowercase separated with `-` instead of camelcase and get on with it. And then I'll add an option to override the path.

Reviewer 2: > Extracting a name from a controller will be trickier because a user can name his classes differently. We could recommend a convention but who knows if they'll follow it.

The point of conventions is that people follow them. If they don't they can't customise the output.

Entities are already singular by convention. Controller names don't have a convention other than the brain-dead "resource" that jax-rs got confused about.

It's pretty easy to tell people to name their controller `People` and default to using that name.

Author: Let's discuss it on https://github.com/quarkusio/quarkus/issues/9261 since the guide change won't be enough for this anyway.

If your application has an entity (e.g. `Person`) that extends either `PanacheEntity` or `PanacheEntityBase` class, you could instruct Panache REST to generate its CRUD resource with the following interface:

[source,java]
----
public interface PersonCrudResource extends PanacheEntityCrudResource<Person, Long> {
}
----

=== PanacheRepositoryCrudResource

If your application has a simple entity (e.g. `Person`) and a repository (e.g. `PersonRepository`) that implements either `PanacheRepository` or `PanacheRepositoryBase` interface, you could instruct Panache REST to generate its CRUD resource with the following interface:

[source,java]
----
public interface PersonCrudResource extends PanacheRepositoryCrudResource<PersonRepository, Person, Long> {
}
----

== Resource customisation

Panache REST provides a `@PanacheRestResource` annotation that can be used to customize certain features of the resource.
It can be used on your CRUD resource interface:

[source,java]
----
@PanacheRestResource(hal = true)
public interface PersonCrudResource extends PanacheEntityCrudResource<Person, Long> {
    @PanacheRestResource(exposed = false)
    void delete(Long id);
}
----

=== Available options

> * `hal` - in addition to the standard `application/json` responses, generates additional methods that can return `application/hal+json` responses if requested via `Accept` header.

Reviewer 3: Wondering if we should have an array of content types instead with the default being `{ "application/json" }`. That would be much more flexible.

Reviewer 3: /cc @FroMage

Reviewer: Well, I don't think we currently support any response type, but this seems like a valid point for the future.

Author: `application/json` is hardcoded and `application/hal+json` can be enabled as an extra. Because HAL is a complex format I had to implement a custom serialiser for it which serialisers special wrapper objects.

Reviewer 3: OK, let's keep it that way for now.
